lib/ansible/executor/play_iterator.py

```diff
@@ -400,6 +439,41 @@ def _get_next_task_from_state(self, state, host):
                             task = None
                         state.cur_always_task += 1
 
+            elif state.run_state == IteratingStates.HANDLERS:
+                if not state._handlers_sorted:
+                    # handlers are executed in the order they are defined, not in the order notified
+                    if state.handlers:
+                        all_handlers = [h for b in self._play.handlers for h in b.block]
+                        state._notified_handlers = deque((h for h in all_handlers if h in state.handlers))
+                    state._handlers_sorted = True
+
+                if state.handlers_child_state:
+                    state.handlers_child_state, task = self._get_next_task_from_state(state.handlers_child_state, host=host)
+                    if self._check_failed_state(state.handlers_child_state):
+                        state.handlers_child_state = None
+                        self._set_failed_state(state)
+                    else:
+                        if task is None or state.handlers_child_state.run_state == IteratingStates.COMPLETE:
+                            state.handlers_child_state = None
+                            continue
+                else:
+                    if state.fail_state != FailedStates.NONE and not self._play.force_handlers:
+                        state._handlers_sorted = False
+                        state.run_state = IteratingStates.COMPLETE
+                    elif len(state.handlers) == 0:
+                        state._handlers_sorted = False
+                        state.run_state = state.pre_flushing_run_state
+                    else:
+                        task = state.pop_current_handler()
```

Comment:
🤯 

This makes me think that using `deque` for notifications complicates things. I'll play with this a bit more and try to revert the notifications to be on handlers as in `devel` which should fix this and eliminate the need to sort the handlers to the defined order. Might make the lockstep easier too.

Thanks!

Reply:
Hrm, I wrote this playbook to reproduce the issue:
```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - command: echo
      notify:
        - handler1
        - handler2
  handlers:
    - name: handler1
      debug:
        msg: handler1
    - name: handler2
      debug:
        msg: handler2
      changed_when: true
      notify:
        handler1
```

but the behavior does not change, it is same in `devel` and with this PR. I think it's because in `devel` we run `run_handlers` after the specific strategy ends running in https://github.com/ansible/ansible/blob/e4890afc4e2af6c5126d056a2b4fdb1f1f122cf6/lib/ansible/plugins/strategy/__init__.py#L322 I thought it was mostly to implement `force_handlers` but it appears it is also to ensure to flush any remaining handlers.

What was your reproducer?

Reply:
So I guess this is resolved! Maybe good to have a test though.

Reply:
Yeah, I wouldn't be surprised if this was fixed during some completely unrelated change.

+1, I'll add a test, this is a good scenario to cover.

Reply:
Oh! I was on the wrong branch. I can still reproduce:
```
- hosts: localhost
  gather_facts: false
  tasks:
    - command: echo
      notify: handler1
  handlers:
    - name: handler1
      command: echo
      notify: handler2
    - name: handler2
      command: echo
      notify: handler1
```

Reply:
That is evil :-)

I have a WIP diff at https://gist.github.com/mkrizek/68aba4e74f74ff5fa34dba41f042af11 that fixes this issue. I am not sure if it's better than what is currently in this PR though. I should at least run `ci_complete` on it.

Reply:
this is a bug, we clearly document you can ONLY notify handlers that have not executed yet

Reply:
though we do seem to have some code that ends up limiting this ...
```
- hosts: localhost
  gather_facts: false
  tasks:
    - command: echo
      notify:
        - h3
  handlers:
    - name: h3
      debug: msg=h3
      changed_when: true
      notify: h4

    - name: h4
      debug: msg=h4
      changed_when: true
      notify: h3
```
causes extra executions (each is run 3 times) but not infinite loop (runs both 2 times if i notify h4 from command task)

Reply:
yeah, my concern was not about keeping backwards compat, just not to introduce infinite recursion. Erroring would be fine with me too.